src/NuGet.Core/NuGet.Packaging/Signing/Verification/SignatureTrustAndValidityVerificationProvider.cs

```diff
@@ -94,9 +94,10 @@ private PackageVerificationResult VerifyValidityAndTrust(Signature signature, Si
             var timestamp = timestamps.FirstOrDefault();
             if (timestamp != null)
             {
-                using (var authorSignatureNativeCms = NativeCms.Decode(signature.SignedCms.Encode(), detached: false))
+                using (var primarySignatureNativeCms = NativeCms.Decode(signature.SignedCms.Encode(), detached: false))
                 {
-                    var signatureHash = NativeCms.GetSignatureValueHash(signature.SignatureContent.HashAlgorithm, authorSignatureNativeCms);
+                    var timestampHashAlgorithmName = CryptoHashUtility.OidToHashAlgorithmName(timestamp.TstInfo.HashAlgorithmId.Value);
```

Comment:
I don't like that we are hashing with a potentially unsupported algorithm, but it will still fail later if the algorithm is not supported. This could go in like this, I'm optimizing this as part of my work in https://github.com/NuGet/NuGet.Client/pull/2006 to check first if the algorithm is supported and then verifying integrity.